### PR TITLE
Feature/clean up topword

### PR DESCRIPTION
**Pull request by Weiqi97**
**_Description_**
* Rewrote get_class_division_map function, it now returns a panda data frame.
  * Under the effect of changing return type, made corresponding changes in top_word_interface.py file to let the function correctly detect the number of existing classes.
  * For the same reason above, made corresponding changes in templates/topword.html to let it correctly read the data.
* Clean up processors/analyze/topword.py
  * Remove everything that related to unused parameter "option"
  * Changed input data type to panda data frame
  * Make corresponding changes in managers/utility.py
  * Add unit test

### CODE REVIEW

File: lexos/helpers/error_messages.py [hunk 1]
```diff
@@ -17,3 +17,4 @@
                             " the length of the given document"
 INVALID_CUTTING_TYPE_MESSAGE = "the cutting type should be letters, lines, " \
     "number, words or milestone"
+EMPTY_NP_ARRAY_MESSAGE = "the numpy array should not be empty"
```
Comment (chantisnake) [inline lexos/helpers/error_messages.py]: This error message is too vague.

File: lexos/managers/file_manager.py [hunk 2]
```diff
@@ -1359,59 +1359,31 @@ def get_matrix_deprec(self, use_word_tokens: bool, use_tfidf: bool,
         return count_matrix
 
     def get_class_division_map(self):
```
Comment (chantisnake) [inline lexos/managers/file_manager.py]: type hinting

File: lexos/managers/file_manager.py [hunk 3]
```diff
@@ -1359,59 +1359,31 @@ def get_matrix_deprec(self, use_word_tokens: bool, use_tfidf: bool,
...
-                        [request.form["file_" + str(files[file_id].id)]])
-                except KeyError:
-                    name_map.append([files[file_id].label])
-                class_label_map.append(files[file_id].class_label)
-
-        return division_map, name_map, class_label_map
+        for file in files:
+            temp_labels.append(file.label)
+            if file.class_label not in temp_classes:
+                temp_classes.append(file.class_label)
```
Comment (chantisnake) [inline lexos/managers/file_manager.py]: 喵喵喵？这也太难看了吧= =

File: lexos/managers/file_manager.py [hunk 4]
```diff
@@ -1359,59 +1359,31 @@ def get_matrix_deprec(self, use_word_tokens: bool, use_tfidf: bool,
...
+
+        # initialize class division map
+        division_map = pd.DataFrame(
+            data=np.zeros((class_length, label_length), dtype=bool),
+            index=temp_classes,
+            columns=temp_labels)
+
+        # set correct boolean value for each file
+        for file in files:
+            division_map[file.label][file.class_label] = True
```
Comment (chantisnake) [inline lexos/managers/file_manager.py]: Okay, but a little unintuitive, can you find a way without initializing the data frame? 

Construct it directly

File: lexos/managers/utility.py [hunk 5]
```diff
@@ -1287,134 +1286,81 @@ def generate_sims_csv(file_manager: FileManager):
 
 
 def get_top_word_option():
```
Comment (chantisnake) [inline lexos/managers/utility.py]: type hinting

File: lexos/managers/utility.py [hunk 6]
```diff
@@ -1287,134 +1286,81 @@ def generate_sims_csv(file_manager: FileManager):
...
         human_result = []
-        for i in range(len(analysis_result)):
-            header = 'Document "' + \
-                     temp_labels[i] + '" compared to the whole corpus'
-            human_result.append([header, analysis_result[i]])
-
-    elif test_by_class == 'classToPara':  # test by class
+        for index, label in enumerate(labels):
+            header = 'Document "' + label + '" compared to the whole corpus'
+            human_result.append([header, analysis_result[index]])
```
Comment (chantisnake) [inline lexos/managers/utility.py]: for loop with append... hum... try panda dataframe or panda series

File: lexos/managers/utility.py [hunk 7]
```diff
@@ -1287,134 +1286,81 @@ def generate_sims_csv(file_manager: FileManager):
...
-            group_word_lists, option=option, low=low, high=high)
+        analysis_result = analyze_para_to_group(group_values=group_values,
+                                                words=words)
 
         # convert to human readable form
         human_result = []
         for key in list(analysis_result.keys()):
             file_name = name_map[key[0]][key[1]]
-            comp_class_name = class_label_map[key[2]]
+            comp_class_name = class_labels[key[2]]
```
Comment (chantisnake) [inline lexos/managers/utility.py]: for loop with append... hum... try panda dataframe or panda series

File: lexos/managers/utility.py [hunk 8]
```diff
@@ -1462,34 +1405,27 @@ def generate_z_test_top_word(file_manager: FileManager):
 
 
 def get_top_word_csv(test_results, csv_header):
```
Comment (chantisnake) [inline lexos/managers/utility.py]: if you change the data structure in `generate_topwords` to data frame, you will be able to directly save them as csv. 
http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.to_csv.html

File: lexos/processors/analyze/topword.py [hunk 9]
```diff
@@ -1,580 +1,249 @@
...
-            # LowIQR: only analyze the Left outlier of word, determined by iqr
-            high = (mid - 1.5 * iqr)
-
-        else:
-            raise IOError('input option is not valid')
-
-    else:
-        raise IOError('input option is not valid')
+    # initialize
+    word_z_score_dict = {}
```
Comment (chantisnake) [inline lexos/processors/analyze/topword.py]: try data frame instead of dict

File: lexos/processors/analyze/topword.py [hunk 10]
```diff
@@ -1,580 +1,248 @@
...
+    assert nt > 0, SEG_NON_POSITIVE_MESSAGE
     p = (p1 * n1 + pt * nt) / (n1 + nt)
-    try:
-        standard_error = sqrt(p * (1 - p) * ((1 / n1) + (1 / nt)))
-        z_scores = (p1 - pt) / standard_error
-        return z_scores
-    except ZeroDivisionError:
-        return 'Insignificant'
+    standard_error = (p * (1 - p) * ((1 / n1) + (1 / nt)))**0.5
+    if standard_error == 0:
```
Comment (chantisnake) [inline lexos/processors/analyze/topword.py]: for float, always use close, not `==`

File: lexos/processors/analyze/topword.py [hunk 11]
```diff
@@ -1,580 +1,248 @@
...
-            low = (mid - 1.5 * iqr)
-
-        elif option.startswith('low'):
-            # LowIQR: only analyze the Left outlier of word, determined by iqr
-            high = (mid - 1.5 * iqr)
-
-        else:
-            raise IOError('input option is not valid')
+    # initialize
+    word_z_score_dict = {}
```
Comment (chantisnake) [inline lexos/processors/analyze/topword.py]: dictionary is not the best data structure here

File: lexos/processors/analyze/topword.py [hunk 12]
```diff
@@ -1,580 +1,248 @@
...
 
+def group_division(dtm: pd.DataFrame, division_map: np.ndarray) -> \
+        (List[np.ndarray], List[np.ndarray]):
+    """Divides the word counts into groups via the group map.
 
-def truncate(x, d):
-    """
-    add this function to keep d digits after decimal point of a number x
+    :param dtm: pandas data frame that contains the word count matrix.
+    :param division_map: a numpy matrix represents the group map.
```
Comment (chantisnake) [inline lexos/processors/analyze/topword.py]: explain in more detail what the `division_map` looks like

File: lexos/processors/analyze/topword.py [hunk 13]
```diff
@@ -1,580 +1,248 @@
...
-                    'Custom')
-
-    :return:    contain a array
-                each element of array is a array, represent a chunk and it is
-                    sorted via z_score
-                each element array is a tuple: (word, corresponding z_score)
+def analyze_all_to_para(count_matrix: np.ndarray,
+                        words: np.ndarray,
+                        labels: np.ndarray) -> List[Tuple[str, list]]:
+    """Analyzes each single word compare to the total documents
```
Comment (chantisnake) [inline lexos/processors/analyze/topword.py]: end with period.

File: lexos/processors/analyze/topword.py [hunk 14]
```diff
@@ -1,580 +1,248 @@
...
-    # there are two element in the tuple, each one is a index of groups
+    # There are two elements in the tuple, each one is a index of groups
     # (for example the first group will have index 0)
-    # two group index cannot be equal
-    comp_map = itertools.product(
-        list(
-            range(num_group)), list(
-            range(num_group)))
+    # Two groups index cannot be equal.
+    comp_map = itertools.product(list(range(num_group)),
```
Comment (chantisnake) [inline lexos/processors/analyze/topword.py]: I am not sure, does `zip` works here?

Comment (Weiqi97) [inline lexos/processors/analyze/topword.py]: I tried and I think maybe zip doesn't really work here...

File: lexos/processors/analyze/topword.py (hunk 11 above)
Comment (Weiqi97) [inline lexos/processors/analyze/topword.py]: Sorry I may need a bit more time to change is.. Sorry about keeping this PR this long

Comment (chantisnake) [inline lexos/processors/analyze/topword.py]: That is totally fine, that is how long pr is supposed to take.

File: lexos/processors/analyze/topword.py [hunk 15]
```diff
@@ -1,580 +1,254 @@
...
+        (List[np.ndarray], List[np.ndarray]):
+    """Divides the word counts into groups via the group map.
+
+    :param dtm: pandas data frame that contains the word count matrix.
+    :param division_map: a numpy matrix represents the group map, where the
+                         index represents the class names and the column
+                         contains file name of each file. The matrix contains
+                         boolean values to determine which class each file
+                         belongs to.
+    :return: a list of lists, each list represents a group, where each element
```
Comment (chantisnake) [inline lexos/processors/analyze/topword.py]: documentation outdated?

**Comment by chantisnake** [comment]
I am still think the data structure used here is not perfect.

Let's talk about this later

**Comment by chantisnake** [comment]
Approve this for now. I will create an issue to discuss the data structure issue

**Pull request merged**